<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Introduction</title>
    <meta charset="utf-8" />
    <meta name="author" content="Bengt Sennblad" />
    <meta name="date" content="2020-11-16" />
    <script src="libs/header-attrs-2.8/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/metropolis.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/metropolis-fonts.css" rel="stylesheet" />
    <script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
    <script src="libs/viz-1.8.2/viz.js"></script>
    <link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet" />
    <script src="libs/grViz-binding-1.0.6.1/grViz.js"></script>
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Introduction
## Machine learning and Biostatistics course
### Bengt Sennblad
### NBIS
### 11/16/2020

---




# Content/aim



.pull-left[
## What do we want you to gain from this course?
- Learn a selection statistical methods

{{content}}
]
--

#### ...but, more importantly, be able to
- by yourselves explore statistics
  - find alternative methods
  - get a intuitive understanding
  - evaluate assumptions
      + know when to apply
      + know when not to apply

--
.pull-right[
### How?
{{content}}
]
--
- Go from &lt;span style="background-color:black"&gt;&lt;span style="color:white"&gt;_black box_&lt;/span&gt;&lt;/span&gt; to understanding
    - several levels
      - models = assumptions
      - tests -- test statistics
      - overfitting
{{content}}
--
- Appreciate theory as a ❤️*friend*❤️
    - help understanding assumptions
    - help when evaluating methods
{{content}}
--
- Provide a _#framework#_ for learning
    - statistical thinking
    - basic math language
      - not afraid of equations
      - know some basic terminology

---
class: inverse, middle, center
&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=800px&gt;&lt;/html&gt; 

# Why Statistics?
## Examples


---

# Example: Mouse knockout experiment

### Setup
Hypothesis: The low density lipoprotein (LDL) receptor gene *LDLR* affects the prevalence of hypercholsterolaemia.

- 10 wildtype (wt) mice
- 10 *LDLR* knockout (KO) mice

Measure plasma concentration of total cholesterol at one time point after feeeding on high fat diet.

???

low-density lipoprotein
---
layout: true
# Example: Mouse knockout experiment
.pull-left[
![](intro2_files/figure-html/mice1-1.png)&lt;!-- --&gt;
]
---
.pull-right[
### Visualize results in a plot

]
---
.pull-right[
### Visualize results in a plot
- Not so informative!
- Let's improve
]
---
layout: false
layout: true
# Example: Mouse knockout experiment

.pull-left[
![](intro2_files/figure-html/mice1a-1.png)&lt;!-- --&gt;
]

---
.pull-right[

### Improved visualization
  - Collect KO and wt separately as columns in a **boxplot**
]

---
.pull-right[

### Improved visualization
  - Collect KO and wt separately as columns in a boxplot
  ![black box warning](assets/blackbox.jpg)
]

---
.pull-right[

### Improved visualization
  - Collect KO and wt separately as columns in a boxplot
  - Visualize **distribution** of sample values using **descriptive statistics**
      - *mean* 
      - *quartiles*
]
---
.pull-right[

### Interpretation

Intuitively, there is a clear difference between KO and wt in this plot...
]
???
- 

-- 
.pull-right[
What if the result had looked different?

]
---
layout: false
# Example: Mouse knockout experiment

.pull-left[
![](intro2_files/figure-html/mice3-1.png)&lt;!-- --&gt;
]
.pull-right[

### Interpretation

...equally clearly, here is no difference between KO and wt...

]
---
layout: true
# Example: Mouse knockout experiment

.pull-left[
![](intro2_files/figure-html/mice4-1.png)&lt;!-- --&gt;
]

---
.pull-right[

### Interpretation

... Now, the difference is less clear

{{content}}
]
--

... we get uncertain

{{content}}

--
... Uncertainty best measured as probabilities!

{{content}}
--

... with probabilities we can perform a **statistical test** for difference

{{content}}
--
  ![black box warning](assets/blackbox.jpg)
]

---
.pull-right[
### Statistical tests

#### Model
- Abstractions of reality  probabilities
- Simplifying assumptions
  - Variation approximately `\(N(\mu,\sigma)\)`
    
#### Hypotheses
- H0: `\(\mu\)` and `\(\sigma\)` same for KO and wt 
- H1: `\(\mu\)` might be different 

#### Test
- **P-value** `\((p)\)` =  probability of the observed or more extreme difference  under H0
- **significance level** `\(p\leq \alpha\)`
]

---
layout: false 
layout: true
# Example: Mouse knockout experiment

.pull-left[
![](intro2_files/figure-html/mice5-1.png)&lt;!-- --&gt;
]
---
.pull-right[

### Interpretation

... what about here?

]


---
.pull-right[

### Interpretation

... what about here?

#### Tentative answer
This might require testing hypotheses that also look at the difference in  *variance*

]


---
layout: false
# Example: Protein expression

#### Setup

**Aim:** Investigate the relation between the Breast cancer transcriptomic and proteomics landscape

- Breast cancer tumour samples experiment
    + measure RNA expression
    + measure protein expression




---

layout: true
# Example: Protein expression

.pull-left[
![](intro2_files/figure-html/regression2-1.png)&lt;!-- --&gt;
]
---
.pull-right[
### Discovery: 
- **Correlation** between Protein and RNA expression

{{content}}
]
--
  ![black box warning](assets/blackbox.jpg)

???

Actually, not going to through this here .. have to wait until the session on this
---

.pull-right[
### Discovery: 
- **Correlation** between Protein and RNA expression

### Hypothesis: 
- RNA expression governs protein expression

{{content}}
]
--

### Test
- **regression**
    + **fit** a **linear model** 
{{content}}

--
  ![black box warning](assets/blackbox.jpg)

---
layout: false
layout: true
# Example: Protein expression

.pull-left[
![](intro2_files/figure-html/regression3-1.png)&lt;!-- --&gt;
]
---
.pull-right[
### Linear model

- Extend our previous model
    + Variation of `\([Prot]\)` approximately `\(N(\mu,\sigma)\)`  
    *and*
    + `\(\mu = \alpha +\beta [RNA]\)` 
{{content}}
]

???
Maybe this can suffice as hierarchical model example instead...?
--
- Uses
  1. Simulation
      - validating methods
{{content}}
--
  2. Inference 
      - **regression**
          - optimal `\(\alpha\)` and `\(\beta\)` values
      - **significance test**
      
???
- `\(\beta\)` tells how strongly y depend on x
- not always this simple
---
layout: false
layout: true
# Example: Protein expression

.pull-left[
![](intro2_files/figure-html/regression4-1.png)&lt;!-- --&gt;
]

---
.pull-right[
### Linear model

- Extend our previous model
    + Variation of `\([Prot]\)` approximately `\(N(\mu,\sigma)\)`  
    *and*
    + `\(\mu = \alpha +\beta [RNA]\)` 
- Uses
  1. Simulation
      - validating methods
  2. Inference 
      - **regression**
          - optimal `\(\alpha\)` and `\(\beta\)` values
      - **significance test**
]


???
More realistic

These researchers found that there were a portion of genes low correlation RNA  and protein expression



---
layout: false
layout: true
# Example: Protein expression

.pull-left[
![](intro2_files/figure-html/regression5-1.png)&lt;!-- --&gt;
]
---
.pull-right[
### Another hypothesis: 
- RNA-protein correlation is governed by RNA variation

### `\(\Leftarrow\)` Result
- How should this be interpreted?

{{content}}
]
--
### Non-linear dependencies
- **Generalized linear models (GLMs)**
{{content}}
--
  ![black box warning](assets/blackbox.jpg)

]

---
layout: false
# Example: GWAS

## Setup: 

Aim: Elucidate the genetic setup of IMD condition

- Case-control cohort (N = 1000)
- Genome-wide SNP assay (P= 5000)



---
layout: true
# Example: GWAS


.pull-left[
![](intro2_files/figure-html/gwas-1.png)&lt;!-- --&gt;

]
---
.pull-right[

## Manhattan plot

- Individual association test for each SNP and disease
- `\(-log_{10}\)` P-value against genomic position
{{content}}
]

???
-log(0.05) = 1.30103 (line)
- several houndred or even thousand associations
--


#### *Multiple tests*
- Repeated experiments are more likely to succeed under H0
{{content}}


???

- Discuss repeated dice throws
- This manhattan plot is actually generated under the null model
--

#### Multiple test correction
- **Bonferroni, False Discovery Rate (FDR)**
{{content}}

--

  ![black box warning](assets/blackbox.jpg)

]


---
layout: false
# Example: Protein expression cntd.


### What other variables affect protein expression?

- How can we investigate that?

---
layout: true
# Example: Protein expression cntd.

.pull-left[
![](intro2_files/figure-html/multivariate-1.png)&lt;!-- --&gt;
]

---
.pull-right[
## Multivariate regression

{{content}} 
]

--

 ![black box warning](assets/blackbox.jpg) 
]

---
.pull-right[
## Multivariate regression
- We can visualize models with 3 variables
- and write an equation for this:  
`\(\begin{cases}[prot] &amp;\sim N(\mu, \sigma) \\\mu = &amp;\alpha+\beta [RNA] \\&amp;\quad + \gamma [proteases] \end{cases}\)`
{{content}}
]

--
- For &gt;3 variables, visualization breaks...
{{content}}
--
- but we can still write it as an equation:  
`\(\begin{cases}[prot] \sim&amp; N(\mu, \sigma) \\\mu =&amp; \alpha+\beta [RNA] \\&amp;\quad+ \gamma [proteases]\\ &amp;\quad + \delta [miRNA]\end{cases}\)`

---
layout: false
# Example: Protein expression cntd.
.pull-left[
![](intro2_files/figure-html/multivariate2-1.png)&lt;!-- --&gt;
]
.pull-right[
#### Issues with multivariate analysis

1. **Feature selection**
    - Include only the most relevant variables in out model
{{content}}
]
--
2. **Overfitting**
    - Risk of modeling random noise in data
    - Solution: **regularization** or **Bayesian approaches**
{{content}}
--
![black box warning](assets/blackbox.jpg) 

{{content}}
--

#### Really Big Data 
- **Machine Learning** on all data
{{content}}


---
layout: false
layout: true
# Example: Single cell expression data

### Unsupervised Machine Learning - Dimensional reduction
.pull-left[
![](intro2_files/figure-html/pca-1.png)&lt;!-- --&gt;
]
---
.pull-right[

### Feature selection with PCA

- **linear transformation** of original variables into new *hidden variables*, components/factors

### Clustering
- identify groups based on hidden variables
]
---
.pull-right[

### Feature selection with PCA

- **linear transformation** of original variables into new *hidden variables*, components/factors

### Clustering
- identify groups based on hidden variables
  ![black box warning](assets/blackbox.jpg)

]
---
layout: false
layout: true

# Example: Breast Cancer classification

## Supervised Machine learning -- classification

- **Train**, e.g., a *Random forest classifier* as a cell-type classifier in single cell RNAseq.

Training data: Breast Cancer Wisconsin Data 

.pull-left[

X: measures on

1. Clump Thickness
2. Uniformity of Cell Size
3. Uniformity of Cell Shape
4. Marginal Adhesion
5. Single Epithelial Cell Size
6. Bare Nuclei
7. Bland Chromatin
8. Normal Nucleoli
9. Mitoses

]
.pull-right[

Y: Breast Cancer class

1. Benign tumour
2. Malign tumour

]

---

<div id="htmlwidget-c824cf028bca3879fc0b" style="width:504px;height:504px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-c824cf028bca3879fc0b">{"x":{"diagram":"digraph {\n\ngraph [layout = dot, rankdir = LR]\nnode [shape = rectangle, style = filled, fillcolor = Linen]\n\n# label nodes\ninpout [label = \"Mass Spec Data\", shape = folder, fillcolor = Beige]\nRF [label = \"Random Forest \n Classifier\", shape = folder, fillcolor = Beige]\noutput [label =  \"Outcome\"]\n\n\ninput -> RF -> output\n\n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script>



---
##### Training input  could be, e.g., images (pixels)


---

#####  ... or training input could be a genome sequence...

---

#####  This is of course a major...
![black box warning](assets/blackbox.jpg)

???
This is of course a major black box

---
layout: false
# Welcome to the course!

&lt;br&gt;
&lt;br&gt;

&lt;center&gt;
![black box warning](assets/blackboxgift.png)
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"slideNumberFormat": "<div class=\"progress-bar-container\">\n  <div class=\"progress-bar\" style=\"width: calc(%current% / %total% * 100%);\">\n  </div>\n</div>"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
