<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Eva Freyhult, Mun-Gwan Hong" />

<meta name="date" content="2022-09-15" />

<title>Exercises: Clustering</title>

<script src="exercises_files/header-attrs-2.16/header-attrs.js"></script>
<link href="exercises_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="exercises_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>










</head>

<body>




<h1 class="title toc-ignore">Exercises: Clustering</h1>
<h4 class="author">Eva Freyhult, Mun-Gwan Hong</h4>
<h4 class="date">2022-09-15</h4>



<div id="exercise-1" class="section level4">
<h4>Exercise 1</h4>
<p>Using Edgar Anderson’s iris data, we will investigate K-means
clustering. The data is stored in <code>iris</code> in R. Here we focus
on <code>Sepal.Width</code> and <code>Petal.Width</code>.</p>
<p><a
href="https://www.publicdomainpictures.net/en/view-image.php?image=342859&amp;picture=iris-flower"><img
src="images/iris-flower-1592163215Z1C.jpg" alt="Iris" /></a></p>
<p> </p>
<ol style="list-style-type: lower-alpha">
<li><p>Examine the contents and data types of the data set
<em>iris</em>.</p></li>
<li><p>Try to find three clusters in the data of sepal and petal widths
using a k-means algorithm. You can use the function <code>kmeans</code>
in R. It allows to choose an algorithm with the argument
<code>algorithm</code> (e.g. <code>algorithm = "Forgy"</code>).</p></li>
<li><p>Plot the data and color by cluster ID</p></li>
<li><p>Repeat the same analysis a few times. Do you get the same
clusters all the time? Use <code>table</code> to check it allowing
different cluster numbers.</p></li>
<li><p>Try the argument <code>nstart</code> out and compare how stable
two runs are. By setting the argument to 4, the algorithm will
automatically try four different (random) starting points.</p></li>
<li><p>Try different values for <span class="math inline">\(k\)</span>,
run the k-means algorithm and collect the WSS
(<code>tot.withinss</code>). Plot WSS vs <span
class="math inline">\(k\)</span>. The WSS is always decreasing as <span
class="math inline">\(k\)</span> increases, but the curve can still give
you a hint of which <span class="math inline">\(k\)</span> to choose.
The Elbow method for selecting <span class="math inline">\(k\)</span> is
to look at this curve and choose the <span
class="math inline">\(k\)</span> that you find at the bend of the curve,
at the ‘elbow’. Which <span class="math inline">\(k\)</span> would you
pick based on this?</p></li>
<li><p>Compute the Silhouette width of the first object (or observation)
when <span class="math inline">\(k\)</span> is 2.</p></li>
<li><p>(Optional) Compute average Silhouette width of all objects
adjusting <span class="math inline">\(k\)</span> from 2 to 10 and choose
optimal <span class="math inline">\(k\)</span>. Euclidean distance can
be computed by <code>dist</code>. Check the help of the function using
<code>?dist</code>.</p></li>
<li><p>(Optional) Try <code>fviz_nbclust</code> of the R-package
<code>factoextra</code> and choose optimal <span
class="math inline">\(k\)</span> based on average Silhouette
width.</p></li>
</ol>
<p> </p>
</div>
<div id="exercise-2" class="section level4">
<h4>Exercise 2</h4>
<p>The NCI60 data set consists of gene expression values for 6830 genes
for 64 cell lines. Using this data set, we investigate a few
hierarchical clustering distances and linkage methods. The data can be
downloaded in R using the following command</p>
<pre><code>nci_data &lt;- read.table(
  url(&quot;https://web.stanford.edu/~hastie/ElemStatLearn/datasets/nci.data.csv&quot;),
  sep = &quot;,&quot;,
  row.names = 1,
  header = TRUE
)
nci_label &lt;- scan(
  url(&quot;https://web.stanford.edu/~hastie/ElemStatLearn/datasets/nci.label.txt&quot;),
  what = &quot;&quot;
)</code></pre>
<p> </p>
<ol style="list-style-type: lower-alpha">
<li><p>What is the size of the data matrix? Do every column represent a
gene or a cell line?</p></li>
<li><p>Compute the Euclidean distance <strong>between cell
lines</strong>. This can be accomplished using the function
<code>dist</code>. Read the help text <code>?dist</code>. This function
computes the distance between rows of the input data matrix. If the rows
represent cell lines, you can run <code>dist(nci_data)</code>. But, if
your cell lines are represented by columns, you need to transpose the
data matrix first <code>t(nci_data)</code>.</p></li>
<li><p>Cluster the cell lines using complete linkage hierarchical
clustering, use the function <code>hclust</code>.</p></li>
<li><p>Plot the dendrogram of the clustering result (The help for
<code>hclust</code> includes the function for the plot)</p>
<ul>
<li>Try changing the labels to something more informative using
<code>nci_label</code>.</li>
<li>Investigate the argument <code>hang</code>, what happens if you set
it to -1?</li>
</ul></li>
<li><p>Try the linkage methods “single”, “average” and “ward.D” in
addition to “complete”. Compare the results. Which method is
‘best’?</p></li>
</ol>
<p>Pick the tree resulting from the method you think is ‘best’. How many
clusters are there?</p>
<p>You can cut the tree on any level to get between 1 and 64 clusters.
The function <code>cutree</code> either on a specific height
(dissimilarity) or to get a specific number of clusters.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
