<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Clustering - 1&nbsp; Clustering: art of finding groups</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./exercises.html" rel="next">
<link href="./index.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="site_libs/quarto-diagram/mermaid.min.js"></script>
<script src="site_libs/quarto-diagram/mermaid-init.js"></script>
<link href="site_libs/quarto-diagram/mermaid.css" rel="stylesheet">

<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./intro.html"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Clustering: art of finding groups</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Clustering</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Clustering: art of finding groups</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercises</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1.1</span> Introduction</a></li>
  <li><a href="#common-usage" id="toc-common-usage" class="nav-link" data-scroll-target="#common-usage"><span class="header-section-number">1.2</span> Common usage</a></li>
  <li><a href="#partition-methods-vs.-hierarchical" id="toc-partition-methods-vs.-hierarchical" class="nav-link" data-scroll-target="#partition-methods-vs.-hierarchical"><span class="header-section-number">1.3</span> Partition methods vs.&nbsp;hierarchical</a></li>
  <li><a href="#distance-and-similarity-measures" id="toc-distance-and-similarity-measures" class="nav-link" data-scroll-target="#distance-and-similarity-measures"><span class="header-section-number">1.4</span> Distance and similarity measures</a>
  <ul>
  <li><a href="#distances-for-interval-variables" id="toc-distances-for-interval-variables" class="nav-link" data-scroll-target="#distances-for-interval-variables"><span class="header-section-number">1.4.1</span> Distances for interval variables</a>
  <ul class="collapse">
  <li><a href="#euclidean-distance" id="toc-euclidean-distance" class="nav-link" data-scroll-target="#euclidean-distance">Euclidean distance</a></li>
  <li><a href="#squared-euclidean-distance" id="toc-squared-euclidean-distance" class="nav-link" data-scroll-target="#squared-euclidean-distance">Squared Euclidean distance</a></li>
  <li><a href="#chebychev" id="toc-chebychev" class="nav-link" data-scroll-target="#chebychev">Chebychev</a></li>
  <li><a href="#city-block-or-manhattan-distance" id="toc-city-block-or-manhattan-distance" class="nav-link" data-scroll-target="#city-block-or-manhattan-distance">City block or Manhattan distance</a></li>
  <li><a href="#mahalanobis-distance" id="toc-mahalanobis-distance" class="nav-link" data-scroll-target="#mahalanobis-distance">Mahalanobis distance</a></li>
  <li><a href="#minkowski" id="toc-minkowski" class="nav-link" data-scroll-target="#minkowski">Minkowski</a></li>
  <li><a href="#canberra" id="toc-canberra" class="nav-link" data-scroll-target="#canberra">Canberra</a></li>
  </ul></li>
  <li><a href="#distances-for-binary-data" id="toc-distances-for-binary-data" class="nav-link" data-scroll-target="#distances-for-binary-data"><span class="header-section-number">1.4.2</span> Distances for binary data</a>
  <ul class="collapse">
  <li><a href="#dice" id="toc-dice" class="nav-link" data-scroll-target="#dice">Dice</a></li>
  <li><a href="#jaccard" id="toc-jaccard" class="nav-link" data-scroll-target="#jaccard">Jaccard</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#partition-methods-k-means-pam" id="toc-partition-methods-k-means-pam" class="nav-link" data-scroll-target="#partition-methods-k-means-pam"><span class="header-section-number">1.5</span> Partition methods: k-means &amp; PAM</a>
  <ul>
  <li><a href="#k-means" id="toc-k-means" class="nav-link" data-scroll-target="#k-means"><span class="header-section-number">1.5.1</span> k-means</a></li>
  <li><a href="#pam-partition-around-medoids" id="toc-pam-partition-around-medoids" class="nav-link" data-scroll-target="#pam-partition-around-medoids"><span class="header-section-number">1.5.2</span> PAM, partition around medoids</a></li>
  </ul></li>
  <li><a href="#hcl-hierarchical-clustering" id="toc-hcl-hierarchical-clustering" class="nav-link" data-scroll-target="#hcl-hierarchical-clustering"><span class="header-section-number">1.6</span> HCL, hierarchical clustering</a>
  <ul>
  <li><a href="#linkage-agglomerative-clustering" id="toc-linkage-agglomerative-clustering" class="nav-link" data-scroll-target="#linkage-agglomerative-clustering"><span class="header-section-number">1.6.1</span> Linkage &amp; Agglomerative clustering</a></li>
  <li><a href="#getting-clusters" id="toc-getting-clusters" class="nav-link" data-scroll-target="#getting-clusters"><span class="header-section-number">1.6.2</span> Getting clusters</a></li>
  </ul></li>
  <li><a href="#how-many-clusters" id="toc-how-many-clusters" class="nav-link" data-scroll-target="#how-many-clusters"><span class="header-section-number">1.7</span> How many clusters?</a>
  <ul>
  <li><a href="#elbow-method" id="toc-elbow-method" class="nav-link" data-scroll-target="#elbow-method"><span class="header-section-number">1.7.1</span> Elbow method</a></li>
  <li><a href="#silhouette-method" id="toc-silhouette-method" class="nav-link" data-scroll-target="#silhouette-method"><span class="header-section-number">1.7.2</span> Silhouette method</a></li>
  <li><a href="#gap" id="toc-gap" class="nav-link" data-scroll-target="#gap"><span class="header-section-number">1.7.3</span> Gap</a></li>
  <li><a href="#pvclust" id="toc-pvclust" class="nav-link" data-scroll-target="#pvclust"><span class="header-section-number">1.7.4</span> pvclust</a></li>
  </ul></li>
  <li><a href="#two-way-clustering" id="toc-two-way-clustering" class="nav-link" data-scroll-target="#two-way-clustering"><span class="header-section-number">1.8</span> Two-way clustering</a></li>
  <li><a href="#additional-comments-resources" id="toc-additional-comments-resources" class="nav-link" data-scroll-target="#additional-comments-resources"><span class="header-section-number">1.9</span> Additional comments &amp; resources</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Clustering: art of finding groups</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dendextend)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(knitr)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(kableExtra)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cluster)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggalt)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(factoextra)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(NbClust)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pvclust)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(circlize)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>dir_images <span class="ot">&lt;-</span> <span class="st">"images/"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="introduction" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">1.1</span> Introduction</h2>
<ul>
<li><p>Clustering is about finding structure in the data by identifying natural groupings (clusters). In other words, clustering aims to group object according to similarity. As a results, <strong>a cluster</strong> is a collection of objects (samples, genes etc.) that are more “similar” to each other that they are to cases in other clusters.</p></li>
<li><p>Clustering for a set of <span class="math inline">\(n\)</span> objects is usually performed based on a selection of <span class="math inline">\(p\)</span> variables. We can for instance cluster <span class="math inline">\(n\)</span> blood samples based on <span class="math inline">\(p\)</span> gene expression measurements or we can cluster <span class="math inline">\(n\)</span> genes based on the <span class="math inline">\(p\)</span> samples blood profiles.</p></li>
<li><p>Clustering is <strong>unsupervised learning</strong>, where we do not use any data labels to perform grouping. It is commonly used in EDA to reveal structures in the data.</p></li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">include_graphics</span>(<span class="st">"images/fruits.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-clust" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="images/fruits.png" class="img-fluid figure-img" width="457"></p>
<figcaption class="figure-caption">Figure&nbsp;1.1: Grouping fruits (A) can be done in more than one natural way. For instance, we can group by fruit type (B) or by fruit color (C). Often, the usefulness of the grouping is in the eye of the beholder, something that should not be forgotten when promoting a particular grouping over other possibilities.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="common-usage" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="common-usage"><span class="header-section-number">1.2</span> Common usage</h2>
<ul>
<li>Grouping samples based on clinical data</li>
<li>Grouping samples based on gene expression profiles</li>
<li>Grouping cells by cell types, e.g.&nbsp;in scRNA-seq</li>
<li>Biomarker studies</li>
<li>Disease subtyping and stratification</li>
<li>Clustering metagenomics contigs</li>
<li>Species classification and biodiversity studies</li>
</ul>
</section>
<section id="partition-methods-vs.-hierarchical" class="level2" data-number="1.3">
<h2 data-number="1.3" class="anchored" data-anchor-id="partition-methods-vs.-hierarchical"><span class="header-section-number">1.3</span> Partition methods vs.&nbsp;hierarchical</h2>
<div class="cell">
<div class="cell-output-display">
<div>
<div>
<pre class="mermaid mermaid-js">flowchart LR
  A([Type of splitting]) -.-&gt; B([Partition])
  A([Type of splitting]) -.-&gt; C([Hierarchical])
  C([Hierarchical]) -.-&gt; D([Agglomerative])
  C([Hierarchical]) -.-&gt; E([Divisive])

</pre>
</div>
</div>
</div>
</div>
<p><br></p>
<ul>
<li><p>The partitioning and hierarchical methods are two general classes of clustering methods.</p></li>
<li><p>In <strong>partitioning</strong> methods address the problem of dividing <span class="math inline">\(n\)</span> objects, described by <span class="math inline">\(p\)</span> variables, into a small number (<span class="math inline">\(k\)</span>) of discrete classes.</p></li>
<li><p>For instance, considering only two genes A and B (<span class="math inline">\(p=2\)</span>) we can group 70 blood samples (<span class="math inline">\(n=70\)</span>) into three clusters (<span class="math inline">\(k=3\)</span>).</p></li>
</ul>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># simulate data</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">190</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">class =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"cl1"</span>, <span class="st">"cl2"</span>, <span class="st">"cl3"</span>), <span class="fu">c</span>(<span class="dv">30</span>, <span class="dv">15</span>, <span class="dv">25</span>))) <span class="sc">|&gt;</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">rnorm</span>(<span class="fu">n</span>(), <span class="fu">c</span>(<span class="at">cl1 =</span> <span class="dv">1</span>, <span class="at">cl2 =</span> <span class="dv">3</span>, <span class="at">cl3 =</span> <span class="dv">2</span>)[class], <span class="at">sd =</span> .<span class="dv">35</span>), </span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">rnorm</span>(<span class="fu">n</span>(), <span class="fu">c</span>(<span class="at">cl1 =</span> <span class="dv">1</span>, <span class="at">cl2 =</span> <span class="dv">1</span>, <span class="at">cl3 =</span> <span class="dv">2</span>)[class], <span class="at">sd =</span> .<span class="dv">35</span>),</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="co"># plot data</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>text_size <span class="ot">&lt;-</span> <span class="dv">12</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>mytheme <span class="ot">&lt;-</span> <span class="fu">theme</span>(<span class="at">legend.title =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>                 <span class="at">legend.position =</span> <span class="st">"top"</span>, </span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>                 <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> text_size),</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>                 <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>                 <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>                 <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> text_size),</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>                 <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> text_size))</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>mycols <span class="ot">&lt;-</span> <span class="fu">brewer.pal</span>(<span class="dv">6</span>, <span class="st">"Set1"</span>)</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df, <span class="fu">aes</span>(x, y)) <span class="sc">+</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span> </span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"gene A"</span>) <span class="sc">+</span> </span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"gene B"</span>) <span class="sc">+</span> </span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>  mytheme </span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a><span class="co"># show clusters</span></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df, <span class="fu">aes</span>(x, y, <span class="at">color =</span> class)) <span class="sc">+</span></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span> </span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"gene A"</span>) <span class="sc">+</span> </span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"gene B"</span>) <span class="sc">+</span> </span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a>  mytheme <span class="sc">+</span> </span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-partition" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="intro_files/figure-html/fig-partition-1.png" class="img-fluid figure-img" width="864"></p>
<figcaption class="figure-caption">Figure&nbsp;1.2: Illustration of partition data splitting. A) Mock-up data for 70 blood samples for which expression of two genes were measured (left). B) Following cluster analysis, samples are grouped into three clusters (right).</figcaption>
</figure>
</div>
</div>
</div>
<p><br></p>
<ul>
<li><strong>Divisive hierarchical</strong> methods begin with all of the objects in one cluster which is broken down into sub-cluster until all objects are separated. <strong>The agglomerative approach</strong> works in the opposite direction as single-member clusters are increasingly fused (joint) until all objects are in one cluster.</li>
</ul>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">include_graphics</span>(<span class="st">"images/hclust-ann.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-hclust-ann" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="images/hclust-ann.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption class="figure-caption">Figure&nbsp;1.3: A dendrogram representation for hierarchical clustering of data objects.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="distance-and-similarity-measures" class="level2" data-number="1.4">
<h2 data-number="1.4" class="anchored" data-anchor-id="distance-and-similarity-measures"><span class="header-section-number">1.4</span> Distance and similarity measures</h2>
<ul>
<li>All clustering algorithms start by measuring the <strong>similarity</strong> between the objects to be clustered. Similar cases will be placed into the same cluster.</li>
<li>It is also possible to view similarity by its inverse, the <strong>distance</strong> between the objects, with distance declining as similarity increases.</li>
<li>There are large number of <strong>distance metrics</strong> by which distance can be measured. Some are restricted to particular data types.</li>
<li>All distance metrics share the property that distance declines with increasing similarity. However, they <strong>do not</strong> share a common distance between the same two cases, i.e.&nbsp;<strong>distance changes with the similarity measure</strong>, leading to potential alternative clustering results.</li>
<li>Three general classes of distance measures are <strong>Euclidean metrics</strong>, <strong>Non-Euclidean metrics</strong> and <strong>Semi-metrics</strong>.</li>
</ul>
<p><strong>Euclidean metrics</strong></p>
<ul>
<li>These metrics measure true straight-line distances in Euclidean space.</li>
<li>In a univariate example the Euclidean distance between two values is the arithmetic difference. In a bivariate case the minimum distance between two points is the hypotenuse of a triangle formed from the points (Pythagoras theorem). For three variables the hypotenuse extends through three-dimensional space. In n-dimensional space, the Euclidean distance between two points <span class="math inline">\(A = (a_1, a_2, ..., a_n)\)</span> and <span class="math inline">\(B = (b_1, b_2, ..., b_n)\)</span></li>
</ul>
<p><span class="math display">\[distance(A,B) = \sqrt{\sum_{i=1}^{n}(a_i - b_i)^2}\]</span></p>
<ul>
<li>For instance, given two objects, red circle at <span class="math inline">\((5, 3)\)</span> and yellow cylinder at <span class="math inline">\((3, 1)\)</span> the Euclidean distance is <span class="math inline">\(d = \sqrt{(3 - 5)^2 + (1 - 3)^2} \approx 2.82\)</span></li>
</ul>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">include_graphics</span>(<span class="st">"images/distances.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-distances" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="images/distances.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption class="figure-caption">Figure&nbsp;1.4: Ilustration of distance metrics for a bivariate case. A) Euclidean metric, a distance between two point is sthe hypotenuse of a triangle formed from the points B) Non-Euclidean Manhatan metric.</figcaption>
</figure>
</div>
</div>
</div>
<p><strong>Non-Euclidean metrics</strong></p>
<ul>
<li>These are distances that are not straight lines, but which obey four rules. Let <span class="math inline">\(d_{ij}\)</span> be the distance between two objects, <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>.</li>
</ul>
<ol type="1">
<li><span class="math inline">\(d_{ij}\)</span> must be 0 or positive (objects are identical, <span class="math inline">\(d_{ij} = 0\)</span>, or they are different, <span class="math inline">\(d_{ij} &gt; 0\)</span>)</li>
<li><span class="math inline">\(d_{ij} = d_{ji}\)</span>, distance from A to B is the same as from B to A</li>
<li><span class="math inline">\(d_{jj} = 0\)</span>, an object is identical to itself</li>
<li><span class="math inline">\(d_{ik} \le d_{ij} + d_{jk}\)</span>, when considering three objects the distance between any two of them cannot exceed the sum of the distances between the other two pairs. In other words, the distances can construct a triangle.</li>
</ol>
<ul>
<li>The <strong>Manhattan</strong> or City Block metric is an example of this type of distance metrics. <span class="math display">\[distance (A,B) = \sum|a_i - b_i|\]</span></li>
<li>In our red circle and yellow cylinder, we get: <span class="math inline">\(d = |3 - 5| + |1 - 3| = 4\)</span></li>
</ul>
<p><strong>Semi metric</strong></p>
<ul>
<li>Semi metric distances obey the first three aforementioned rules, but may not obey the last “triangle rule”.</li>
<li>The <strong>cosine</strong> measure is an example of semi metric distance class. It is often used to compare vectors in text mining, where documents are often represented as vectors of numbers that represent word/term frequencies.</li>
</ul>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">include_graphics</span>(<span class="st">"images/cosine.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-cosine" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="images/cosine.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption">Figure&nbsp;1.5: Ilustration of cosine distance metric, used for comparing vectors and text queries.</figcaption>
</figure>
</div>
</div>
</div>
<section id="distances-for-interval-variables" class="level3" data-number="1.4.1">
<h3 data-number="1.4.1" class="anchored" data-anchor-id="distances-for-interval-variables"><span class="header-section-number">1.4.1</span> Distances for interval variables</h3>
<section id="euclidean-distance" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="euclidean-distance">Euclidean distance</h4>
<ul>
<li>Already mentioned above, normal Pythagoras theorem extended to the appropriate number of dimensions.</li>
<li>Can be used when measuring similarity between points in Euclidean space, e.g.&nbsp;working with spatial data such as geographic coordinates or points on a map.</li>
<li>It may not be the best choice for high-dimensional data due to “distance concentration” phenomena. In high-dimensional space, the pair-wise distances between different data points converge to the same value as the dimensionality increases, making it hard to distinguish points that are near and far away.</li>
</ul>
</section>
<section id="squared-euclidean-distance" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="squared-euclidean-distance">Squared Euclidean distance</h4>
<ul>
<li>The sum of the squared differences between scores for two cases on all variables, i.e.&nbsp;the squared length of the hypotenuse. This measure magnifies distances between cases that are further apart.</li>
<li>Note that the squared Euclidean distance is not a metric as it does not fulfill the triangle inequality.</li>
</ul>
</section>
<section id="chebychev" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="chebychev">Chebychev</h4>
<ul>
<li><p>The absolute maximum difference, on one variable, between two cases. <span class="math display">\[distance(A, B) = max |a_i - b_i|\]</span></p></li>
<li><p>This measures examines distances across all of the variables, but only uses the maximum distance. This one dimensional difference need not be constant across cases. For example, age could be used for one pair and height for another.</p></li>
<li><p>Interesting application of the distance is by <span class="citation" data-cites="Ghaderyan2020">Ghaderyan and Beyrami (<a href="#ref-Ghaderyan2020" role="doc-biblioref">2020</a>)</span> who tried many distance metric in analysis of gait rhythm fluctuations for automatic neurodegenerative disease detection. They achieved good results using Chebyshev and cosine distance to distinguish patients with Amyotrophic Lateral Sclerosis, Parkinson’s disease (PD) and and Huntington’s disease (HD).</p></li>
</ul>
</section>
<section id="city-block-or-manhattan-distance" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="city-block-or-manhattan-distance">City block or Manhattan distance</h4>
<ul>
<li>A distance that follows a route along the non-hypotenuse sides of a triangle. The name refers to the grid-like layout of most American cities which makes it impossible to go directly between two points.</li>
<li>This metric is hence less affected by outliers than the Euclidean and squared Euclidean distances. <span class="math display">\[distance (A,B) = \sum|a_i - b_i|\]</span></li>
<li>Is preferred over Euclidean distance when there is a high dimensionality in the data.</li>
<li>May not be the best choice, if one wants to perform space rotation or wants a smooth and differentiable function (converge more easily).</li>
</ul>
</section>
<section id="mahalanobis-distance" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="mahalanobis-distance">Mahalanobis distance</h4>
<ul>
<li>It is a generalized version of a Euclidean distance which weights variables using the <strong>sample variance-covariance matrix</strong>. Because the covariance matrix is used this also means that correlations between variables are taken into account. <span class="math display">\[distance(A,B) = [(a_i - b_i)^t S^{-1}(a_i - b_i)]^{1/2}\]</span></li>
</ul>
<p>where <span class="math inline">\(S^{-1}\)</span> is the inverse covariance matrix$</p>
</section>
<section id="minkowski" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="minkowski">Minkowski</h4>
<ul>
<li>Minkowski distance is a generalization of both the Euclidean distance and the Manhattan distance in a normed vector space.</li>
<li>It can be thought of as a way of measuring the length of the path between two points when moving along axes aligned with the coordinate system.</li>
<li>The formula for the Minkowski distance of order <span class="math inline">\(p\)</span> between two points <span class="math inline">\(A = (a_1, a_2, ..., a_n)\)</span> and <span class="math inline">\(B = (b_1, b_2, ..., b_n)\)</span> in an <span class="math inline">\(n\)</span>-dimensional space is given by <span class="math display">\[distance(A,B) = \left(\sum_{i=1}^{n} |a_i - b_i|^p\right)^{1/p}\]</span> where <span class="math inline">\(p\)</span> determines the form of the distance: for <span class="math inline">\(p=1\)</span>, it becomes the Manhattan distance (sum of the absolute differences of their coordinates); for <span class="math inline">\(p=2\)</span>, it yields the Euclidean distance (the shortest direct line between two points); and for <span class="math inline">\(p=\infty\)</span>, it approaches the Chebyshev distance (the maximum difference along any coordinate dimension).</li>
<li>The versatility of the Minkowski distance makes it applicable across various fields and data types, allowing adjustments for different spatial concepts and scales.</li>
</ul>
</section>
<section id="canberra" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="canberra">Canberra</h4>
<ul>
<li>The Canberra is a weighted version of the Manhattan distance, as it calculates the absolute difference between two vectors and normalizes it by dividing it by the absolute sum of their values <span class="math display">\[distance(A, B) = \sum_{i=1}^{n} \frac{|a_i - b_i|}{|a_i| + |b_i|}\]</span></li>
<li><span class="citation" data-cites="Blanco2023">Blanco-Mallo et al. (<a href="#ref-Blanco2023" role="doc-biblioref">2023</a>)</span> compared a series of distance metrics for machine learning purposes and showed that Canberra distance had the best overall performance and the highest tolerance to noise.</li>
</ul>
</section>
</section>
<section id="distances-for-binary-data" class="level3" data-number="1.4.2">
<h3 data-number="1.4.2" class="anchored" data-anchor-id="distances-for-binary-data"><span class="header-section-number">1.4.2</span> Distances for binary data</h3>
<ul>
<li>There is large number of metrics available for binary data, and many have more than one name, with Dice and Jaccard measures used probably most often in biology. All of the distance measures for binary data use two or three values obtained from a simple two by two matrix of agreement.</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data =</span> <span class="fu">c</span>(<span class="st">""</span>, <span class="st">""</span>, <span class="st">"Case"</span>, <span class="st">"i"</span>, </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>                     <span class="st">""</span>, <span class="st">""</span>, <span class="st">"+"</span>, <span class="st">"-"</span>, </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"Case j"</span>, <span class="st">"+"</span>, <span class="st">"a"</span>, <span class="st">"b"</span>, </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>                     <span class="st">""</span>, <span class="st">"-"</span>, <span class="st">"c"</span>, <span class="st">"d"</span>), <span class="at">ncol =</span> <span class="dv">4</span>, <span class="at">nrow =</span> <span class="dv">4</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>x <span class="sc">%&gt;%</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> <span class="co">#print()</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kbl</span>(<span class="at">col.names =</span> <span class="cn">NULL</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">align =</span> <span class="fu">c</span>(<span class="st">"c"</span>, <span class="st">"c"</span>, <span class="st">"c"</span>, <span class="st">"c"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable_paper</span>(<span class="at">full_width =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="tbl-binary" class="anchored">
<table class="lightable-paper table table-sm table-striped small" data-quarto-postprocess="true">
<caption>Table&nbsp;1.1: Binary metrics use two or more of the values obtained from a two by two matrix of agrreement.</caption>
<tbody>
<tr class="odd">
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">Case</td>
<td style="text-align: center;">i</td>
</tr>
<tr class="even">
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">+</td>
<td style="text-align: center;">-</td>
</tr>
<tr class="odd">
<td style="text-align: center;">Case j</td>
<td style="text-align: center;">+</td>
<td style="text-align: center;">a</td>
<td style="text-align: center;">b</td>
</tr>
<tr class="even">
<td style="text-align: center;"></td>
<td style="text-align: center;">-</td>
<td style="text-align: center;">c</td>
<td style="text-align: center;">d</td>
</tr>
</tbody>
</table>
</div>


</div>
</div>
<p>where:</p>
<ul>
<li><span class="math inline">\(a\)</span> is the number of cases which both share the attribute</li>
<li><span class="math inline">\(d\)</span> number of cases which neither have the attribute</li>
<li><span class="math inline">\(b\)</span> and <span class="math inline">\(c\)</span> dare the number of cases in which only one of the pair has the attribute.</li>
<li>Note: <span class="math inline">\(a + b + c + d = n\)</span>, the sample size</li>
</ul>
<section id="dice" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="dice">Dice</h4>
<ul>
<li>A similarity measure in which joint absences <span class="math inline">\(d\)</span> are excluded from consideration, and matches <span class="math inline">\(a\)</span> are weighted double.</li>
<li>Also known as Czekanowski or Sorensen measure <span class="math display">\[distance(A, B) = 2a / (2a + b + c)\]</span></li>
</ul>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">include_graphics</span>(<span class="st">"images/dice.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-dice" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="images/dice.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption">Figure&nbsp;1.6: Ilustration of dice distance metric, used for comparing binary data.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="jaccard" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="jaccard">Jaccard</h4>
<ul>
<li>A similarity measure in which joint absences <span class="math inline">\(d\)</span> are excluded from consideration.</li>
<li>Equal weight is given to matches and mismatches.</li>
</ul>
<p><span class="math display">\[distance(A, B) = a / (a + b + c)\]</span></p>
</section>
</section>
</section>
<section id="partition-methods-k-means-pam" class="level2" data-number="1.5">
<h2 data-number="1.5" class="anchored" data-anchor-id="partition-methods-k-means-pam"><span class="header-section-number">1.5</span> Partition methods: k-means &amp; PAM</h2>
<section id="k-means" class="level3" data-number="1.5.1">
<h3 data-number="1.5.1" class="anchored" data-anchor-id="k-means"><span class="header-section-number">1.5.1</span> k-means</h3>
<ul>
<li>The k-means clustering aims to divide all <span class="math inline">\(n\)</span> objects, described by <span class="math inline">\(p\)</span> variables, into exactly <span class="math inline">\(k\)</span> clusters of discrete classes.</li>
<li>The value of <span class="math inline">\(k\)</span>, i.e.&nbsp;number of clusters, has to be given to the algorithm.</li>
<li>The algorithm minimizes the variance within clusters, by iteratively assigning each object to the cluster with the closest <strong>centroid</strong>, where centroid of cluster <span class="math inline">\(k\)</span>th is the arithmetic mean of all <span class="math inline">\(n_k\)</span> objects in the cluster.</li>
</ul>
<p><span class="math display">\[{\mathbf m}_k = \frac{1}{n_k} \sum_{i=1}^{n_k} {\mathbf x}_{i}\]</span></p>
<ul>
<li>The simplest version of k-means algorithm is by Lloyd and Forgy’s <span class="citation" data-cites="Forgy1965">Lloyd (<a href="#ref-Lloyd1982" role="doc-biblioref">1982</a>)</span>. In R, <code>kmeans</code> functions implements a modified more effective version of this algorithm by <span class="citation" data-cites="Hartigan1979">Hartigan and Wong (<a href="#ref-Hartigan1979" role="doc-biblioref">1979</a>)</span>.</li>
</ul>
<p><strong>The Lloyd and Forgy’s algorithm</strong></p>
<ol type="1">
<li><p>Initialization. Select <span class="math inline">\(k\)</span> initial centroids. The initial centroids can be selected in several different ways. Two common methods are</p>
<ul>
<li>Select <span class="math inline">\(k\)</span> data points as initial centroids.</li>
<li>Randomly assign each data point to one out of <span class="math inline">\(k\)</span> clusters and compute the centroids for these initial clusters.</li>
</ul></li>
<li><p>Assign each object to the closest centroid (in terms of squared Euclidean distance). The squared Euclidean distance between an object (a data point) and a cluster centroid <span class="math inline">\(m_k\)</span> is <span class="math inline">\(d_i = \sum_{j=1}^p (x_{ij} - m_{kj})^2\)</span>. By assigning each object to closest centroid the total within cluster sum of squares (WSS) is minimized. <span class="math display">\[WSS = \sum_{k=1}^K\sum_{i \in C_k}\sum_{j=1}^p (x_{ij} - m_{kj})^2\]</span></p></li>
<li><p>Update the centroids for each of the <span class="math inline">\(k\)</span> clusters by computing the centroid for all the objects in each of the clusters.</p></li>
<li><p>Repeat steps 2 and 3 until convergence.</p></li>
</ol>
<div id="fig-kmeans" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="https://hckr.pl/k-means-visualization/"><img src="images/kmeans.png" class="img-fluid figure-img"></a></p>
<figcaption class="figure-caption">kmeans</figcaption>
</figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;1.7: Interactive visualization of k-means algorithm can be found <a href="https://hckr.pl/k-means-visualization/">here</a>.</figcaption>
</figure>
</div>
<p><strong>When to use and not to use k-means</strong></p>
<ul>
<li>k-means is relatively fast and scalable, making it a good choice for large data sets. It works well with continuous, numeric data since it relies on Euclidean distances. Results are easy to interpret due to the simple assignment of data points to clusters. It works good if one knows the number of clusters or can get that the number of clusters for internal or external validation.</li>
<li>On the downside, k-means is sensitive to the initial placement of cluster centroids. It can sometimes converge to local minima, resulting in suboptimal cluster assignments. Outliers can also significantly impact the position of centroids and lead to poor cluster assignments.</li>
</ul>
</section>
<section id="pam-partition-around-medoids" class="level3" data-number="1.5.2">
<h3 data-number="1.5.2" class="anchored" data-anchor-id="pam-partition-around-medoids"><span class="header-section-number">1.5.2</span> PAM, partition around medoids</h3>
<ul>
<li>PAM is very similar to k-means, but instead of using centroids we use <strong>medoids</strong> to represent clusters.</li>
<li><strong>Medoid</strong> is centrally located objects within the cluster.</li>
<li>This makes PAM more robust when compare to k-means, but with higher computational complexity.</li>
</ul>
<p>The algorithm can be described in few steps:</p>
<ul>
<li>Initialization: randomly select <span class="math inline">\(k\)</span> objects as the initial medoids.</li>
<li>Assignment: assign each object to the nearest medoid, forming <span class="math inline">\(k\)</span> clusters.</li>
<li>Update: for each cluster, find the new medoid by selecting the object that minimizes the sum of distances to all other objects in the cluster.</li>
<li>Check for convergence: if none of the medoids have changed, the algorithm has converged. Otherwise, return to step 2 (Assignment) and repeat the process.</li>
</ul>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># example of running k-means and PAM in R (k = 3)</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># simulate data</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">190</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">class =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"cl1"</span>, <span class="st">"cl2"</span>, <span class="st">"cl3"</span>), <span class="fu">c</span>(<span class="dv">30</span>, <span class="dv">15</span>, <span class="dv">25</span>))) <span class="sc">|&gt;</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">rnorm</span>(<span class="fu">n</span>(), <span class="fu">c</span>(<span class="at">cl1 =</span> <span class="dv">1</span>, <span class="at">cl2 =</span> <span class="dv">3</span>, <span class="at">cl3 =</span> <span class="dv">2</span>)[class], <span class="at">sd =</span> .<span class="dv">35</span>), </span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">rnorm</span>(<span class="fu">n</span>(), <span class="fu">c</span>(<span class="at">cl1 =</span> <span class="dv">1</span>, <span class="at">cl2 =</span> <span class="dv">1</span>, <span class="at">cl3 =</span> <span class="dv">2</span>)[class], <span class="at">sd =</span> .<span class="dv">35</span>),</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="co"># extract x and y coordinates, convert to matrix</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(x, y) <span class="sc">%&gt;%</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>() <span class="sc">%&gt;%</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">round</span>(<span class="dv">2</span>)</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a><span class="co"># run k-means and PAM</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>res_kmeans <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(x, <span class="at">centers =</span> <span class="dv">3</span>)</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>res_pam <span class="ot">&lt;-</span> <span class="fu">pam</span>(x, <span class="at">k =</span> <span class="dv">3</span>)</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a><span class="co"># prepare data frames for plotting</span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>df_kmeans <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cluster =</span> res_kmeans<span class="sc">$</span>cluster) <span class="sc">%&gt;%</span></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cluster =</span> <span class="fu">as.factor</span>(cluster))</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>df_pam <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span></span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cluster =</span> res_pam<span class="sc">$</span>clustering) <span class="sc">%&gt;%</span> </span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cluster =</span> <span class="fu">as.factor</span>(cluster))</span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a>df_centroids <span class="ot">&lt;-</span> res_kmeans<span class="sc">$</span>centers <span class="sc">%&gt;%</span> </span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cluster =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cluster =</span> <span class="fu">as.factor</span>(cluster))</span>
<span id="cb9-34"><a href="#cb9-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-35"><a href="#cb9-35" aria-hidden="true" tabindex="-1"></a>df_medoids <span class="ot">&lt;-</span> res_pam<span class="sc">$</span>medoids <span class="sc">%&gt;%</span></span>
<span id="cb9-36"><a href="#cb9-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb9-37"><a href="#cb9-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cluster =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb9-38"><a href="#cb9-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cluster =</span> <span class="fu">as.factor</span>(cluster))</span>
<span id="cb9-39"><a href="#cb9-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-40"><a href="#cb9-40" aria-hidden="true" tabindex="-1"></a><span class="co"># points with k-means clusters and centroids</span></span>
<span id="cb9-41"><a href="#cb9-41" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> df_kmeans <span class="sc">%&gt;%</span></span>
<span id="cb9-42"><a href="#cb9-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">color =</span> cluster)) <span class="sc">+</span> </span>
<span id="cb9-43"><a href="#cb9-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span> </span>
<span id="cb9-44"><a href="#cb9-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> df_centroids, <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size =</span> <span class="dv">5</span>, <span class="at">pch =</span> <span class="dv">10</span>) <span class="sc">+</span> </span>
<span id="cb9-45"><a href="#cb9-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb9-46"><a href="#cb9-46" aria-hidden="true" tabindex="-1"></a>  mytheme <span class="sc">+</span> </span>
<span id="cb9-47"><a href="#cb9-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"gene A"</span>) <span class="sc">+</span> </span>
<span id="cb9-48"><a href="#cb9-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"gene B"</span>) <span class="sc">+</span> </span>
<span id="cb9-49"><a href="#cb9-49" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span>
<span id="cb9-50"><a href="#cb9-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-51"><a href="#cb9-51" aria-hidden="true" tabindex="-1"></a><span class="co"># points with pam clusters and medoids</span></span>
<span id="cb9-52"><a href="#cb9-52" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> df_pam <span class="sc">%&gt;%</span></span>
<span id="cb9-53"><a href="#cb9-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">color =</span> cluster)) <span class="sc">+</span> </span>
<span id="cb9-54"><a href="#cb9-54" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span> </span>
<span id="cb9-55"><a href="#cb9-55" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> df_medoids, <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size =</span> <span class="dv">5</span>, <span class="at">pch =</span> <span class="dv">21</span>) <span class="sc">+</span> </span>
<span id="cb9-56"><a href="#cb9-56" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb9-57"><a href="#cb9-57" aria-hidden="true" tabindex="-1"></a>  mytheme <span class="sc">+</span> </span>
<span id="cb9-58"><a href="#cb9-58" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"gene A"</span>) <span class="sc">+</span> </span>
<span id="cb9-59"><a href="#cb9-59" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"gene B"</span>) <span class="sc">+</span> </span>
<span id="cb9-60"><a href="#cb9-60" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span>
<span id="cb9-61"><a href="#cb9-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-62"><a href="#cb9-62" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-kmeans-pam" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="intro_files/figure-html/fig-kmeans-pam-1.png" class="img-fluid figure-img" width="864"></p>
<figcaption class="figure-caption">Figure&nbsp;1.8: Ilustration of k-means clusters with centroids marked (left) and PAM clusters with medoids marked (right) for k = 3.</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="hcl-hierarchical-clustering" class="level2" data-number="1.6">
<h2 data-number="1.6" class="anchored" data-anchor-id="hcl-hierarchical-clustering"><span class="header-section-number">1.6</span> HCL, hierarchical clustering</h2>
<ul>
<li>Hierarchical clustering does not require the number of clusters to be specified. Instead of creating a single set of clusters it creates a hierarchy of clusterings based on pairwise dissimilarities.</li>
<li>This hierarchy of clusters is often represented as a <strong>dendrogram</strong>.</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">19</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">x =</span> <span class="fu">rnorm</span>(<span class="dv">10</span>), <span class="at">y =</span> <span class="fu">rnorm</span>(<span class="dv">10</span>))</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>h <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(df))</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>dend <span class="ot">&lt;-</span> <span class="fu">as.dendrogram</span>(h) <span class="sc">%&gt;%</span> <span class="fu">set</span>(<span class="st">"branches_lwd"</span>, <span class="dv">3</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(dend)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-dend" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="intro_files/figure-html/fig-dend-1.png" class="img-fluid figure-img" width="1152"></p>
<figcaption class="figure-caption">Figure&nbsp;1.9: Hierarchical clustering creates a hierarchy of clustering based on pairwise dissimilarites, that are often represented as dendrograms.</figcaption>
</figure>
</div>
</div>
</div>
<ul>
<li>There are two strategies for hierarchical clustering <em>agglomerative</em> (bottom-up) and <em>divisive</em> (top-down).</li>
<li>The agglomerative strategy starts at the bottom with all objects in separate clusters and at each level merge a pair of clusters.</li>
<li>The merged pair of clusters are those with the smallest dissimilarity.</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># illustrate HCL agglomerative</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>df_cl <span class="ot">&lt;-</span> <span class="fu">map_dfr</span>(</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,   <span class="co"># step-by-step tree cuts</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  \(ii) df <span class="sc">|&gt;</span> </span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">step =</span> ii, </span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>           <span class="at">Cluster =</span> <span class="fu">as.character</span>(<span class="fu">cutree</span>(h, <span class="dv">11</span> <span class="sc">-</span> ii))) <span class="sc">|&gt;</span> </span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(Cluster) <span class="sc">|&gt;</span> </span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">is_dup =</span> <span class="fu">n</span>() <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">|&gt;</span> </span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ungroup</span>()</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_cl, <span class="fu">aes</span>(x, y, <span class="at">color =</span> Cluster)) <span class="sc">+</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>  ggalt<span class="sc">::</span><span class="fu">geom_encircle</span>(<span class="at">data =</span> <span class="fu">filter</span>(df_cl, is_dup)) <span class="sc">+</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>step, <span class="at">nrow =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>  mytheme <span class="sc">+</span> </span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-hcl-agg" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="intro_files/figure-html/fig-hcl-agg-1.png" class="img-fluid figure-img" width="1152"></p>
<figcaption class="figure-caption">Figure&nbsp;1.10: Ilustration of hierarchical clustering agglomerative (bottom-up) approach.</figcaption>
</figure>
</div>
</div>
</div>
<ul>
<li>The divisive strategy starts at the top with all objects in a single cluster and at each level one cluster is split into two.</li>
<li>The split is chosen to produce the two groups with the largest possible dissimilarity.</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># illustrate HCL divisive</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> cluster<span class="sc">::</span><span class="fu">diana</span>(<span class="fu">dist</span>(df))</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>df_cl <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>df_cl0 <span class="ot">&lt;-</span> pre <span class="ot">&lt;-</span> <span class="fu">mutate</span>(df, <span class="at">step =</span> <span class="dv">10</span>, <span class="at">Cluster =</span> <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>), <span class="at">is_dup =</span> <span class="cn">FALSE</span>)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(ii <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>) {</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  i_merge <span class="ot">&lt;-</span> d<span class="sc">$</span>merge[ii, ]  <span class="co"># output from divisive h-clustering</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  cl <span class="ot">&lt;-</span> pre<span class="sc">$</span>Cluster   <span class="co"># copy previous</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  cl[cl <span class="sc">%in%</span> i_merge] <span class="ot">&lt;-</span> ii</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  nxt <span class="ot">&lt;-</span> <span class="fu">mutate</span>(df, <span class="at">step =</span> <span class="dv">10</span> <span class="sc">-</span> ii, <span class="at">Cluster =</span> cl)</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>  df_cl <span class="ot">&lt;-</span> <span class="fu">c</span>(df_cl, <span class="fu">list</span>(pre <span class="ot">&lt;-</span> nxt))</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>df_cl <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">list</span>(df_cl0), df_cl) <span class="sc">|&gt;</span> </span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">map_dfr</span>(</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>    <span class="sc">~</span>.x <span class="sc">|&gt;</span> </span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mutate</span>(<span class="at">Cluster =</span> <span class="fu">as.integer</span>(<span class="fu">factor</span>(Cluster))) <span class="sc">|&gt;</span>  <span class="co"># drop disappeared cluster number</span></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>      <span class="co"># encircle</span></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>      <span class="fu">group_by</span>(Cluster) <span class="sc">|&gt;</span> </span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>      <span class="fu">mutate</span>(<span class="at">is_dup =</span> <span class="fu">n</span>() <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">|&gt;</span> </span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ungroup</span>()</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span> </span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Cluster =</span> <span class="fu">as.character</span>(Cluster))</span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_cl, <span class="fu">aes</span>(x, y, <span class="at">color =</span> Cluster)) <span class="sc">+</span></span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>  ggalt<span class="sc">::</span><span class="fu">geom_encircle</span>(<span class="at">data =</span> <span class="fu">filter</span>(df_cl, is_dup)) <span class="sc">+</span></span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>step, <span class="at">nrow =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a>  mytheme <span class="sc">+</span> </span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-hcl-divisive" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="intro_files/figure-html/fig-hcl-divisive-1.png" class="img-fluid figure-img" width="1152"></p>
<figcaption class="figure-caption">Figure&nbsp;1.11: Ilustration of hierarchical clustering divsive (top-down) approach.</figcaption>
</figure>
</div>
</div>
</div>
<ul>
<li>With <span class="math inline">\(n\)</span> objects to cluster both strategies will produce a dendrogram representing the <span class="math inline">\(n-1\)</span> levels in the hierarchy.</li>
<li>Each level represent a specific clustering of the objects into disjoint clusters.</li>
<li>The heights of the branches in the dendrogram are proportional to the dissimilarity of the merged/split clusters.</li>
<li>The divisive approach is used less often due to the computational complexity. Here, we will focus more on the commonly used agglomerative approach.</li>
</ul>
<section id="linkage-agglomerative-clustering" class="level3" data-number="1.6.1">
<h3 data-number="1.6.1" class="anchored" data-anchor-id="linkage-agglomerative-clustering"><span class="header-section-number">1.6.1</span> Linkage &amp; Agglomerative clustering</h3>
<ul>
<li>Agglomerative clustering starts with all objects in separate clusters and at each level merge the pair of clusters with the smallest dissimilarity. The pairwise dissimilarities between objects can be computed according to the distance measures discussed above.</li>
<li>We need one more ingredient for hierarchical clustering, that is a method for computing dissimilarity between clusters, known as <strong>linkage method</strong>. There are several linkage methods, as illustrated below.</li>
</ul>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">include_graphics</span>(<span class="st">"images/linkage.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-linkage" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="images/linkage.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption class="figure-caption">Figure&nbsp;1.12: Ilustration of linkage methods to calcualte dissimilarities between clusters.</figcaption>
</figure>
</div>
</div>
</div>
<p>More formally, the dissimilarity between two clusters A and B with objects <span class="math inline">\(a_1, \dots, a_{n_A}\)</span> and <span class="math inline">\(b_1, \dots, b_{n_B}\)</span> are defined as follows.</p>
<p><strong>Single linkage</strong></p>
<ul>
<li>Single linkage takes as a cluster dissimilarity the distance between the two closest objects in the two clusters. <span class="math display">\[d_{sl}(A, B) = \min_{i,j} d(a_i, b_j)\]</span></li>
</ul>
<p><strong>Complete linkage</strong></p>
<ul>
<li>Complete linkage takes as a cluster dissimilarity the distance between the two objects furthest apart in the two clusters. <span class="math display">\[d_{cl}(A, B) = \max_{i,j} d(a_i, b_j)\]</span></li>
</ul>
<p><strong>Average linkage</strong></p>
<ul>
<li>Average linkage takes as a cluster dissimilarity the average distance between the objects in the two clusters. <span class="math display">\[d_{al}(A, B) = \frac{1}{n_A n_B}\sum_i\sum_j d(a_i, b_j)\]</span></li>
</ul>
<p><strong>Ward’s linkage</strong></p>
<ul>
<li><p>Ward’s linkage method minimize the within variance, by merging clusters with the minimum increase in within sum of squares. <span class="math display">\[d_{wl}(A, B) = \sum_{i=1}^{n_A} (a_i - m_{A \cup B})^2 + \sum_{i=1}^{n_B} (b_i - m_{A \cup B})^2 - \sum_{i=1}^{n_A} (a_i - m_{A})^2 - \sum_{i=1}^{n_B} (b_i - m_{B})^2\]</span> where <span class="math inline">\(m_A, m_B, m_{A \cup B}\)</span> are the center of the clusters <span class="math inline">\(A\)</span>, <span class="math inline">\(B\)</span> and <span class="math inline">\(A \cup B\)</span>, respectively.</p></li>
<li><p><em>Note</em> that Ward’s linkage method should not be combined with any dissimilarity matrix as it is based on the squared Euclidean distance. In the R function <code>hclust</code> either the Euclidean or squared Euclidean distance can be used in combination with the linkage <code>method='ward.D2'</code> or <code>method='ward.D</code>, respectively.</p></li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">19</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">x =</span> <span class="fu">rnorm</span>(<span class="dv">10</span>), <span class="at">y =</span> <span class="fu">rnorm</span>(<span class="dv">10</span>)) <span class="co"># simulate data</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">dist</span>(df) <span class="co"># calculate distance</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>h_single <span class="ot">&lt;-</span> <span class="fu">hclust</span>(d, <span class="at">method =</span> <span class="st">"single"</span>)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>h_average <span class="ot">&lt;-</span> <span class="fu">hclust</span>(d, <span class="at">method =</span> <span class="st">"average"</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>h_complete <span class="ot">&lt;-</span> <span class="fu">hclust</span>(d, <span class="at">method =</span> <span class="st">"complete"</span>)</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>h_ward <span class="ot">&lt;-</span> <span class="fu">hclust</span>(d, <span class="at">method =</span> <span class="st">"ward.D2"</span>)</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="do">## Alternatively</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="do">## h_ward &lt;- hclust(d^2, method = "ward.D")</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>dend_single <span class="ot">&lt;-</span> <span class="fu">as.dendrogram</span>(h_single) <span class="sc">%&gt;%</span> <span class="fu">set</span>(<span class="st">"branches_lwd"</span>, <span class="dv">3</span>)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>dend_average <span class="ot">&lt;-</span> <span class="fu">as.dendrogram</span>(h_average) <span class="sc">%&gt;%</span> <span class="fu">set</span>(<span class="st">"branches_lwd"</span>, <span class="dv">3</span>)</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>dend_complete <span class="ot">&lt;-</span> <span class="fu">as.dendrogram</span>(h_complete) <span class="sc">%&gt;%</span> <span class="fu">set</span>(<span class="st">"branches_lwd"</span>, <span class="dv">3</span>)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>dend_ward <span class="ot">&lt;-</span> <span class="fu">as.dendrogram</span>(h_ward) <span class="sc">%&gt;%</span> <span class="fu">set</span>(<span class="st">"branches_lwd"</span>, <span class="dv">3</span>)</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(dend_single, <span class="at">main =</span> <span class="st">"single"</span>)</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(dend_average, <span class="at">main =</span> <span class="st">"average"</span>)</span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(dend_complete, <span class="at">main =</span> <span class="st">"complete"</span>)</span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(dend_ward, <span class="at">main =</span> <span class="st">"Ward's"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-dend-linkage" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="intro_files/figure-html/fig-dend-linkage-1.png" class="img-fluid figure-img" width="1152"></p>
<figcaption class="figure-caption">Figure&nbsp;1.13: HCL agglomerative clustering with different linkage methods used.</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compare dendrograms side-by-side and highlight differences</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dend_diff</span>(dend_single, dend_complete)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-dend-linkage-side1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="intro_files/figure-html/fig-dend-linkage-side1-1.png" class="img-fluid figure-img" width="1152"></p>
<figcaption class="figure-caption">Figure&nbsp;1.14: Side-by-side comparison of dendrograms with single (left) and complete linkage (right).</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compare dendrograms side-by-side and highlight differences</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dend_diff</span>(dend_average, dend_ward)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-dend-linkage-side2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="intro_files/figure-html/fig-dend-linkage-side2-1.png" class="img-fluid figure-img" width="1152"></p>
<figcaption class="figure-caption">Figure&nbsp;1.15: Side-by-side comparison of dendrograms with average (left) and Ward’s linkage (right).</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="getting-clusters" class="level3" data-number="1.6.2">
<h3 data-number="1.6.2" class="anchored" data-anchor-id="getting-clusters"><span class="header-section-number">1.6.2</span> Getting clusters</h3>
<p>To find concrete clusters based on HCL we can cut the tree either using height or using number of clusters.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>dend_complete <span class="sc">%&gt;%</span> <span class="fu">set</span>(<span class="st">"labels_cex"</span>, <span class="dv">2</span>) <span class="sc">%&gt;%</span> <span class="fu">set</span>(<span class="st">"labels_col"</span>, <span class="at">value =</span> <span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">4</span>), <span class="at">k=</span><span class="dv">2</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>   <span class="fu">plot</span>(<span class="at">main =</span> <span class="st">"Two clusters"</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>dend_complete <span class="sc">%&gt;%</span> <span class="fu">set</span>(<span class="st">"labels_cex"</span>, <span class="dv">2</span>) <span class="sc">%&gt;%</span> <span class="fu">set</span>(<span class="st">"labels_col"</span>, <span class="at">value =</span> <span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>), <span class="at">k=</span><span class="dv">3</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>   <span class="fu">plot</span>(<span class="at">main =</span> <span class="st">"Three clusters"</span>)</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="fl">1.6</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-dend-cut" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="intro_files/figure-html/fig-dend-cut-1.png" class="img-fluid figure-img" width="1152"></p>
<figcaption class="figure-caption">Figure&nbsp;1.16: Example of cutting dendrogram, complete linkage, using height into two and three clusters.</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="how-many-clusters" class="level2" data-number="1.7">
<h2 data-number="1.7" class="anchored" data-anchor-id="how-many-clusters"><span class="header-section-number">1.7</span> How many clusters?</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/clust-no.png" class="img-fluid figure-img"></p>
</figure>
</div>
<ul>
<li>As we have seen above clustering is subjective, and there is more than one way to think about groupings of objects. In addition, there are many clustering algorithms to choose from, each with its own set of parameters and assumptions. We need some ways to be able to say something about how good our clustering, or how many clusters we have.</li>
<li>We will focus on <strong>internal validation</strong> that evaluates clustering quality and estimates cluster count without external information. In <strong>external validation</strong> one compares clustering outcomes to known results, measuring label match to select the best algorithm for a dataset.</li>
<li>Internal validation methods evaluate clustering often by measuring the <strong>compactness</strong> and <strong>separation</strong> of the clusters - objects within a cluster exhibit maximal similarity (compactness), while those in separate clusters maintain a high degree of distinction (separation).</li>
<li>There are many ways to measure this, and some common methods include elbow method based on the total sum of squares <span class="math inline">\(WSS\)</span>, Silhouette method or Gap statistic.</li>
<li><code>NbClust</code> package in R implements 30 indexes for determining the optimal number of clusters in a data set to offer the best clustering scheme.</li>
</ul>
<section id="elbow-method" class="level3" data-number="1.7.1">
<h3 data-number="1.7.1" class="anchored" data-anchor-id="elbow-method"><span class="header-section-number">1.7.1</span> Elbow method</h3>
<ul>
<li>The Elbow method is based on the total within sum of squares, <span class="math inline">\(WSS\)</span>, that the k-means algorithm tries to minimize.</li>
<li>By running the algorithm with several different values of <span class="math inline">\(k\)</span>, e.g.&nbsp;from 1 to 10, we can plot <span class="math inline">\(WSS\)</span> as a function of <span class="math inline">\(k\)</span>.</li>
<li>The inflection (bend, elbow) on the curve indicates an optimal number of clusters.</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># simulate data</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">190</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">class =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"cl1"</span>, <span class="st">"cl2"</span>, <span class="st">"cl3"</span>), <span class="fu">c</span>(<span class="dv">30</span>, <span class="dv">15</span>, <span class="dv">25</span>))) <span class="sc">|&gt;</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">rnorm</span>(<span class="fu">n</span>(), <span class="fu">c</span>(<span class="at">cl1 =</span> <span class="dv">1</span>, <span class="at">cl2 =</span> <span class="dv">3</span>, <span class="at">cl3 =</span> <span class="dv">2</span>)[class], <span class="at">sd =</span> .<span class="dv">35</span>), </span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="fu">rnorm</span>(<span class="fu">n</span>(), <span class="fu">c</span>(<span class="at">cl1 =</span> <span class="dv">1</span>, <span class="at">cl2 =</span> <span class="dv">1</span>, <span class="at">cl3 =</span> <span class="dv">2</span>)[class], <span class="at">sd =</span> .<span class="dv">35</span>),</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>factoextra<span class="sc">::</span><span class="fu">fviz_nbclust</span>(df[, <span class="fu">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>)], <span class="at">FUNcluster =</span> kmeans, <span class="at">method=</span><span class="st">"wss"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-elbow" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="intro_files/figure-html/fig-elbow-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Figure&nbsp;1.17: WSS as a function of the number of clusters. The inflection for k = 3 indicates that this is the best number of clusters for the data set.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="silhouette-method" class="level3" data-number="1.7.2">
<h3 data-number="1.7.2" class="anchored" data-anchor-id="silhouette-method"><span class="header-section-number">1.7.2</span> Silhouette method</h3>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/sil.png" class="img-fluid figure-img"></p>
</figure>
</div>
<ul>
<li><p>The silhouette value for a single object <span class="math inline">\(i\)</span> is a value between -1 and 1 that measures how similar the object is to other objects in the same cluster as compared to how similar it is to objects in other clusters.</p></li>
<li><p>The average silhouette over all objects is a measure of how good the clustering is, the higher the value the better is the clustering.</p></li>
<li><p>The silhouette value, <span class="math inline">\(s(i)\)</span>, for an object <span class="math inline">\(i\)</span> in cluster <span class="math inline">\(C_a\)</span> is calculated as follows:</p>
<ol type="1">
<li>Average distance between <span class="math inline">\(i\)</span> and all other objects in the same cluster <span class="math inline">\(C_a\)</span> <span class="math display">\[a(i) = \frac{1}{|C_a|-1} \sum_{j \neq i, j \in C_a} d(i, j)\]</span></li>
<li>Average distance between <span class="math inline">\(i\)</span> and all objects in another cluster <span class="math inline">\(C_b\)</span>, <span class="math inline">\(d(i,C_b)\)</span> and define the minimum; <span class="math display">\[b(i) = \min_{b \neq a} d(i, C_b)\]</span></li>
<li>The Silhouette index is defined as; <span class="math display">\[s(i) = \frac{b(i) - a(i)}{max(a(i), b(i))}\]</span></li>
</ol></li>
<li><p>A silhouette value close to 1 means that the object is very well clustered, a value close to 0 means that the object lies between two clusters. A negative silhouette value means that the object is incorrectly clustered.</p></li>
<li><p>The <strong>average silhouette over all objects</strong> is a measure of how good the clustering is, the higher the value the better is the clustering.</p></li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>factoextra<span class="sc">::</span><span class="fu">fviz_nbclust</span>(df[, <span class="fu">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>)], <span class="at">FUNcluster =</span> kmeans, <span class="at">method=</span><span class="st">"silhouette"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-sil" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="intro_files/figure-html/fig-sil-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Figure&nbsp;1.18: Average silhouette width as a function of the number of clusters. The maxium value for k = 3 indicates that this is the best number of clusters for the data set.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="gap" class="level3" data-number="1.7.3">
<h3 data-number="1.7.3" class="anchored" data-anchor-id="gap"><span class="header-section-number">1.7.3</span> Gap</h3>
<ul>
<li>The gap statistic compares the total within intra-cluster variation for different values of <span class="math inline">\(k\)</span> with their expected values under null reference distribution of the data. The estimate of the optimal clusters will be value that maximize the gap statistic (i.e., that yields the largest gap statistic). This means that the clustering structure is far away from the random uniform distribution of points.</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>factoextra<span class="sc">::</span><span class="fu">fviz_nbclust</span>(df[, <span class="fu">c</span>(<span class="st">"x"</span>, <span class="st">"y"</span>)], <span class="at">FUNcluster =</span> kmeans, <span class="at">method=</span><span class="st">"gap_stat"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-gap" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="intro_files/figure-html/fig-gap-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Figure&nbsp;1.19: Gap statistics as a function of the number of clusters. The maxium value for k = 3 indicates that this is the best number of clusters for the data set.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="pvclust" class="level3" data-number="1.7.4">
<h3 data-number="1.7.4" class="anchored" data-anchor-id="pvclust"><span class="header-section-number">1.7.4</span> pvclust</h3>
<ul>
<li><code>Pvclust</code> R package to assess the uncertainty in hierarchical cluster analysis <span class="citation" data-cites="Suzuki2006">(<a href="#ref-Suzuki2006" role="doc-biblioref">Suzuki and Shimodaira 2006</a>)</span></li>
<li>Pvclust calculates probability values (p-values) for each cluster using bootstrap resampling techniques.</li>
<li>Two types of p-values are available: approximately unbiased (AU) p-value and bootstrap probability (BP) value. Multiscale bootstrap resampling is used for the calculation of AU p-value, which has superiority in bias over BP value calculated by the ordinary bootstrap resampling.</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">include_graphics</span>(<span class="st">"images/pvclust.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-pvclust" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="images/pvclust.png" class="img-fluid figure-img" style="width:100.0%"></p>
<figcaption class="figure-caption">Figure&nbsp;1.20: Bootstrapping for hierarchical clustering with pvclust. Clusters with AU &gt; 95% indicated by by the rectangles are considered to be strongly supported by data.</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="two-way-clustering" class="level2" data-number="1.8">
<h2 data-number="1.8" class="anchored" data-anchor-id="two-way-clustering"><span class="header-section-number">1.8</span> Two-way clustering</h2>
<ul>
<li>When dealing with omics data we often perform two-way clustering, where there is simultaneous clustering, typically hierarchical, of the rows (e.g.&nbsp;gene expression) and columns (e.g.&nbsp;samples).</li>
<li>This is often a default option when showing data on the heatmaps.</li>
<li>We suggest checking out <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/index.html">ComplexHeatmap</a> package that provides rich functionalities for customizing, including splitting columns and rows by clustering solutions</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"data/tissuesGeneExpression.rda"</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">80</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>idx.keep <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(e), n, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>data_genes <span class="ot">&lt;-</span> e[, idx.keep]</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>data_samples <span class="ot">&lt;-</span> tab[idx.keep, ]</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate variance</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>genes_var <span class="ot">&lt;-</span> <span class="fu">apply</span>(data_genes, <span class="dv">1</span>, var)</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>o <span class="ot">&lt;-</span> <span class="fu">order</span>(genes_var, <span class="at">decreasing =</span> T)</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a><span class="co"># select top genes with highest variance</span></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>n_ft <span class="ot">&lt;-</span> <span class="dv">40</span></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> data_genes[o[<span class="dv">1</span><span class="sc">:</span>n_ft], ]</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a><span class="co"># define function to calculate z scores</span></span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>func_cal_z_score <span class="ot">&lt;-</span> <span class="cf">function</span>(x){</span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>  (x <span class="sc">-</span> <span class="fu">mean</span>(x)) <span class="sc">/</span> <span class="fu">sd</span>(x)</span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a>data_zscore <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(x, <span class="dv">1</span>, func_cal_z_score))</span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a>data_heatmap <span class="ot">&lt;-</span> data_zscore</span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a>cols_wong <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#0072b2"</span>, <span class="st">"#e69f00"</span>)</span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-27"><a href="#cb22-27" aria-hidden="true" tabindex="-1"></a><span class="co"># set color scale</span></span>
<span id="cb22-28"><a href="#cb22-28" aria-hidden="true" tabindex="-1"></a>col_fun <span class="ot">=</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="fu">c</span>(cols_wong[<span class="dv">1</span>], <span class="st">"white"</span>, cols_wong[<span class="dv">2</span>]))</span>
<span id="cb22-29"><a href="#cb22-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-30"><a href="#cb22-30" aria-hidden="true" tabindex="-1"></a><span class="co"># annotate columns</span></span>
<span id="cb22-31"><a href="#cb22-31" aria-hidden="true" tabindex="-1"></a><span class="co"># annotations: columns</span></span>
<span id="cb22-32"><a href="#cb22-32" aria-hidden="true" tabindex="-1"></a>ha_col <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(<span class="at">tissue =</span> data_samples<span class="sc">$</span>Tissue,</span>
<span id="cb22-33"><a href="#cb22-33" aria-hidden="true" tabindex="-1"></a>                            <span class="at">col =</span> <span class="fu">list</span>(<span class="at">tissue =</span> <span class="fu">c</span>(<span class="st">"kidney"</span> <span class="ot">=</span> <span class="st">"darkgreen"</span>,</span>
<span id="cb22-34"><a href="#cb22-34" aria-hidden="true" tabindex="-1"></a>                                               <span class="st">"hippocampus"</span> <span class="ot">=</span> <span class="st">"red"</span>,</span>
<span id="cb22-35"><a href="#cb22-35" aria-hidden="true" tabindex="-1"></a>                                               <span class="st">"colon"</span> <span class="ot">=</span> <span class="st">"blue"</span>,</span>
<span id="cb22-36"><a href="#cb22-36" aria-hidden="true" tabindex="-1"></a>                                               <span class="st">"cerebellum"</span> <span class="ot">=</span> <span class="st">"orange"</span>, </span>
<span id="cb22-37"><a href="#cb22-37" aria-hidden="true" tabindex="-1"></a>                                               <span class="st">"liver"</span> <span class="ot">=</span> <span class="st">"yellow"</span>, </span>
<span id="cb22-38"><a href="#cb22-38" aria-hidden="true" tabindex="-1"></a>                                               <span class="st">"placenta"</span> <span class="ot">=</span> <span class="st">"pink"</span>, </span>
<span id="cb22-39"><a href="#cb22-39" aria-hidden="true" tabindex="-1"></a>                                               <span class="st">"endometrium"</span> <span class="ot">=</span> <span class="st">"black"</span>)))</span>
<span id="cb22-40"><a href="#cb22-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-41"><a href="#cb22-41" aria-hidden="true" tabindex="-1"></a><span class="fu">Heatmap</span>(data_heatmap,</span>
<span id="cb22-42"><a href="#cb22-42" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> col_fun,</span>
<span id="cb22-43"><a href="#cb22-43" aria-hidden="true" tabindex="-1"></a>        <span class="at">top_annotation =</span> ha_col,</span>
<span id="cb22-44"><a href="#cb22-44" aria-hidden="true" tabindex="-1"></a>        <span class="at">name =</span> <span class="st">"h1"</span>,</span>
<span id="cb22-45"><a href="#cb22-45" aria-hidden="true" tabindex="-1"></a>        <span class="at">show_row_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb22-46"><a href="#cb22-46" aria-hidden="true" tabindex="-1"></a>        <span class="at">show_column_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb22-47"><a href="#cb22-47" aria-hidden="true" tabindex="-1"></a>        <span class="at">show_row_dend =</span> <span class="cn">TRUE</span>,</span>
<span id="cb22-48"><a href="#cb22-48" aria-hidden="true" tabindex="-1"></a>        <span class="at">show_column_dend =</span> <span class="cn">TRUE</span>, </span>
<span id="cb22-49"><a href="#cb22-49" aria-hidden="true" tabindex="-1"></a>        <span class="at">column_dend_height =</span> <span class="fu">unit</span>(<span class="dv">3</span>, <span class="st">"cm"</span>), </span>
<span id="cb22-50"><a href="#cb22-50" aria-hidden="true" tabindex="-1"></a>    <span class="at">row_dend_width =</span> <span class="fu">unit</span>(<span class="dv">3</span>, <span class="st">"cm"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-heatmap" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="intro_files/figure-html/fig-heatmap-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Figure&nbsp;1.21: Example of a heatmap based on gene expression data collected for seven tissues. Two-way clustering of both genes and samples shows a good seperation between the tissues.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="additional-comments-resources" class="level2" data-number="1.9">
<h2 data-number="1.9" class="anchored" data-anchor-id="additional-comments-resources"><span class="header-section-number">1.9</span> Additional comments &amp; resources</h2>
<ul>
<li>In addition to partitioning and hierarchical clustering methods there is a wide range of other algorithms. Common ones include model-, density, and grid-based methods. Some of them are explained in <a href="https://docs.google.com/presentation/d/1OgFOM7Zo9rQRiXB0r7EBuLg7gFIgaDwYd3pY2EwxWlg/edit?usp=sharing">this presentation</a>. Here, you can also find more details about <strong>external validation</strong> of cluster analysis.</li>
<li>Although historically it was not recommended to apply cluster analysis to <strong>mixed data types</strong> (e.g.&nbsp;both binary and continuous), some distance metrics have been developed, e.g.&nbsp;Gower’s general coefficient to calculate similarity between different data types. Lately, the principles of decision trees have been also used for computing yet another metric for mixed data types, <strong>unsupervised extra trees dissimilarity, UET</strong>. For an interesting study comparing clustering methods for heterogeneous data, where more details about Grower’s and UET can be found, see <span class="citation" data-cites="Preud2021">Preud’homme et al. (<a href="#ref-Preud2021" role="doc-biblioref">2021</a>)</span>.</li>
<li>For a more applied reading in health research, reviewing distance metrics and clustering algorithms see <span class="citation" data-cites="Gao2023">Gao et al. (<a href="#ref-Gao2023" role="doc-biblioref">2023</a>)</span>.</li>
<li>In scRNA-seq analysis it is common to use clustering based on the graph data. In particular, more details about Leiden algorithms can be found <a href="https://www.sc-best-practices.org/cellular_structure/clustering.html">here</a> or in the <span class="citation" data-cites="Traag2019">(<a href="#ref-Traag2019" role="doc-biblioref">Traag, Waltman, and Eck 2019</a>)</span>.</li>
</ul>
</section>
<section id="references" class="level1 unnumbered">
<h1 class="unnumbered">References</h1>


<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-Blanco2023" class="csl-entry" role="listitem">
Blanco-Mallo, Eva, Laura Morán-Fernández, Beatriz Remeseiro, and Verónica Bolón-Canedo. 2023. <span>“Do All Roads Lead to Rome? Studying Distance Measures in the Context of Machine Learning.”</span> <em>Pattern Recognition</em> 141. <a href="https://doi.org/10.1016/j.patcog.2023.109646">https://doi.org/10.1016/j.patcog.2023.109646</a>.
</div>
<div id="ref-Forgy1965" class="csl-entry" role="listitem">
Forgy, E. 1965. <span>“Cluster Analysis of Multivariate Data: Efficiency Vs. Interpretability of Classifications.”</span> <em>Biometrics</em> 21.
</div>
<div id="ref-Gao2023" class="csl-entry" role="listitem">
Gao, Caroline X., Dominic Dwyer, Ye Zhu, Catherine L. Smith, Lan Du, Kate M. Filia, Johanna Bayer, et al. 2023. <span>“An Overview of Clustering Methods with Guidelines for Application in Mental Health Research.”</span> <em>Psychiatry Research</em>. <a href="https://doi.org/10.1016/j.psychres.2023.115265">https://doi.org/10.1016/j.psychres.2023.115265</a>.
</div>
<div id="ref-Ghaderyan2020" class="csl-entry" role="listitem">
Ghaderyan, Peyvand, and Seyede Marziyeh Ghoreshi Beyrami. 2020. <span>“Neurodegenerative Diseases Detection Using Distance Metrics and Sparse Coding: A New Perspective on Gait Symmetric Features.”</span> <em>Computers in Biology and Medicine</em> 120. <a href="https://doi.org/10.1016/j.compbiomed.2020.103736">https://doi.org/10.1016/j.compbiomed.2020.103736</a>.
</div>
<div id="ref-Hartigan1979" class="csl-entry" role="listitem">
Hartigan, J. A., and M. A. Wong. 1979. <span>“Algorithm AS 136: A k-Means Clustering Algorithm.”</span> <em>Applied Statistics</em> 28. <a href="https://doi.org/10.2307/2346830">https://doi.org/10.2307/2346830</a>.
</div>
<div id="ref-Lloyd1982" class="csl-entry" role="listitem">
Lloyd, Stuart P. 1982. <span>“Least Squares Quantization in PCM.”</span> <em>IEEE Transactions on Information Theory</em> 28. <a href="https://doi.org/10.1109/TIT.1982.1056489">https://doi.org/10.1109/TIT.1982.1056489</a>.
</div>
<div id="ref-Preud2021" class="csl-entry" role="listitem">
Preud’homme, Gregoire, Kevin Duarte, Kevin Dalleau, Claire Lacomblez, Emmanuel Bresso, Malika Smaïl-Tabbone, Miguel Couceiro, et al. 2021. <span>“Head-to-Head Comparison of Clustering Methods for Heterogeneous Data: A Simulation-Driven Benchmark.”</span> <em>Scientific Reports</em> 11. <a href="https://doi.org/10.1038/s41598-021-83340-8">https://doi.org/10.1038/s41598-021-83340-8</a>.
</div>
<div id="ref-Suzuki2006" class="csl-entry" role="listitem">
Suzuki, Ryota, and Hidetoshi Shimodaira. 2006. <span>“Pvclust: An r Package for Assessing the Uncertainty in Hierarchical Clustering.”</span> <em>Bioinformatics</em> 22. <a href="https://doi.org/10.1093/bioinformatics/btl117">https://doi.org/10.1093/bioinformatics/btl117</a>.
</div>
<div id="ref-Traag2019" class="csl-entry" role="listitem">
Traag, V. A., L. Waltman, and N. J. van Eck. 2019. <span>“From Louvain to Leiden: Guaranteeing Well-Connected Communities.”</span> <em>Scientific Reports</em> 9. <a href="https://doi.org/10.1038/s41598-019-41695-z">https://doi.org/10.1038/s41598-019-41695-z</a>.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./index.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Preface</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./exercises.html" class="pagination-link">
        <span class="nav-page-text">Exercises</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>